tracer.go

```diff
@@ -0,0 +1,279 @@
+	if !trace.SpanFromContext(ctx).IsRecording() {
+		return ctx
+	}
+
+	opts := []trace.SpanStartOption{
+		trace.WithSpanKind(trace.SpanKindClient),
+		trace.WithAttributes(t.attrs...),
+		trace.WithAttributes(attribute.String("db.table", data.TableName.Sanitize())),
+	}
+
+	if conn != nil && conn.Config() != nil {
+		opts = append(opts,
+			trace.WithAttributes(attribute.String("database.host", conn.Config().Host)),
+			trace.WithAttributes(attribute.Int("database.port", int(conn.Config().Port))),
+			trace.WithAttributes(attribute.String("database.user", conn.Config().User)))
+	}
+
+	ctx, _ = t.tracer.Start(ctx, "copy_from "+data.TableName.Sanitize(), opts...)
+
+	return ctx
+}
+
+// TraceCopyFromEnd is called at the end of CopyFrom calls.
+func (t *Tracer) TraceCopyFromEnd(ctx context.Context, _ *pgx.Conn, data pgx.TraceCopyFromEndData) {
+	span := trace.SpanFromContext(ctx)
+	recordError(span, data.Err)
+
+	span.End()
+}
+
+// TraceBatchStart is called at the beginning of SendBatch calls. The returned
+// context is used for the rest of the call and will be passed to
+// TraceBatchQuery and TraceBatchEnd.
+func (t *Tracer) TraceBatchStart(ctx context.Context, conn *pgx.Conn, data pgx.TraceBatchStartData) context.Context {
+	if !trace.SpanFromContext(ctx).IsRecording() {
+		return ctx
+	}
+
+	var size int
+	if b := data.Batch; b != nil {
+		size = b.Len()
+	}
+
+	opts := []trace.SpanStartOption{
+		trace.WithSpanKind(trace.SpanKindClient),
+		trace.WithAttributes(t.attrs...),
+		trace.WithAttributes(attribute.Int("pgx.batch.size", size)),
+	}
+
+	if conn != nil && conn.Config() != nil {
+		opts = append(opts,
+			trace.WithAttributes(attribute.String("database.host", conn.Config().Host)),
+			trace.WithAttributes(attribute.Int("database.port", int(conn.Config().Port))),
+			trace.WithAttributes(attribute.String("database.user", conn.Config().User)))
+	}
+
+	ctx, _ = t.tracer.Start(ctx, "batch start", opts...)
+
+	return ctx
+}
+
+// TraceBatchQuery is called at the after each query in a batch.
+func (t *Tracer) TraceBatchQuery(ctx context.Context, conn *pgx.Conn, data pgx.TraceBatchQueryData) {
+	opts := []trace.SpanStartOption{
+		trace.WithSpanKind(trace.SpanKindClient),
+		trace.WithAttributes(t.attrs...),
+	}
+
+	if conn != nil && conn.Config() != nil {
+		opts = append(opts,
+			trace.WithAttributes(attribute.String("database.host", conn.Config().Host)),
+			trace.WithAttributes(attribute.Int("database.port", int(conn.Config().Port))),
+			trace.WithAttributes(attribute.String("database.user", conn.Config().User)))
+	}
+
+	if t.logSQLStatement {
+		opts = append(opts, trace.WithAttributes(semconv.DBStatementKey.String(data.SQL)))
+	}
+
+	spanName := "batch query " + data.SQL
+	if t.trimQuerySpanName {
+		spanName = "query " + strings.Split(data.SQL, " ")[0]
+	}
+
+	_, span := t.tracer.Start(ctx, spanName, opts...)
+	recordError(span, data.Err)
+
+	span.End()
```

Comment:
does it really make sense to have a span just to record the error? aren't we already recording this error in the batch start/end?

Reply:
To be honest: I don't know 🙈 
The `SendBatch` abstraction is pgx-only, where queries will be batched into a transaction. I'm not 100% sure how to properly model this. For example batch end won't have the SQL statements so errors there will be without context. Since we don't use this feature of pgx at the moment, I would suggest just leaving it as it is and see if we get any improvements or bug reports from the community. WDYT?

Reply:
ok, sounds good